File: solutions/python/backtracking/restore-ip-addresses.py
# LeetCode 93 — Restore IP Addresses (Medium)
# Category: Backtracking · Approach: Backtracking
# Time: O(1) | Space: O(1)
# Source: https://leetcode.com/problems/restore-ip-addresses/

def restore_ip_addresses(s: str) -> list[str]:
    res: list[str] = []
    n = len(s)

    def valid(seg: str) -> bool:
        if len(seg) == 1:
            return True
        return seg[0] != "0" and int(seg) <= 255

    def dfs(start: int, parts: list[str]) -> None:
        if len(parts) == 4:
            if start == n:
                res.append(".".join(parts))
            return
        for length in range(1, 4):
            if start + length > n:
                break
            seg = s[start:start + length]
            if valid(seg):
                dfs(start + length, parts + [seg])

    dfs(0, [])
    return res

File: solutions/typescript/backtracking/restore-ip-addresses.ts
// LeetCode 93 — Restore IP Addresses (Medium)
// Category: Backtracking · Approach: Backtracking
// Time: O(1) | Space: O(1)
// Source: https://leetcode.com/problems/restore-ip-addresses/

function restoreIpAddresses(s: string): string[] {
  const res: string[] = [];
  const n = s.length;
  const valid = (seg: string): boolean =>
    seg.length === 1 || (seg[0] !== "0" && Number(seg) <= 255);
  const dfs = (start: number, parts: string[]): void => {
    if (parts.length === 4) {
      if (start === n) res.push(parts.join("."));
      return;
    }
    for (let len = 1; len <= 3 && start + len <= n; len++) {
      const seg = s.slice(start, start + len);
      if (valid(seg)) dfs(start + len, [...parts, seg]);
    }
  };
  dfs(0, []);
  return res;
}
